### PR TITLE
[Process] Windows: stronger escaping, disabled delayed expansion, deprecated options

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | yes
| Tests pass?   | yes
| Fixed tickets | #19993, #10486 
| License       | MIT
| Doc PR        | -

Related to #3381.

### CODE REVIEW

**Comment by nicolas-grekas** [comment]
Creadits to @johnstevenson

**Comment by dunglas** [comment]
Maybe should you copy the full test suite too? https://github.com/johnstevenson/winbox-args/blob/master/tests/ArgsTest.php

**Comment by nicolas-grekas** [comment]
@johnstevenson could you please have a look at this PR, and see if the current failures+patch seems legit to you? Also, OK for adding the test suite here?

**Comment by johnstevenson** [comment]
@nicolas-grekas Will do. No worries copying the test suite

**Comment by staabm** [comment]
@johnstevenson did you check whether this 'advanced' windows escaping logic should be fixed/applied to php-src?

**Comment by nicolas-grekas** [comment]
See https://blogs.msdn.microsoft.com/twistylittlepassagesallalike/2011/04/23/everyone-quotes-command-line-arguments-the-wrong-way/ also for reference. Looks like the escaping logic should vary if bypass_shell is on/off.

**Comment by johnstevenson** [comment]
Yes, that's what the `meta` param is all about.

**Comment by nicolas-grekas** [comment]
PR green
Status: needs review

Changelog updated:
 * disabled delayed expansion on Windows
 * added second argument to ProcessUtils::escapeArgument() for controlling the escaping context on Windows
 * deprecated `!VAR!` expansion allowed by ProcessUtils::escapeArgument()
 * deprecated configuring `proc_open()` options
 * deprecated configuring enhanced Windows compatibility
 * deprecated configuring enhanced SIGCHLD compatibility

One could consider some of these changes as breaking BC. Yet, any other options/uses of the related default behaviors was putting the component in some "unknown/unpredictable" state. It could work really.

Eg delayed expansion what enabled as a workaround to allow using `!env_vars!` in assetic. Yet, delayed expansion adds yet another parsing layer on top of the already too many and clumsy ones on Windows, making argument escaping impossible.
Now that we have `%env(var)%` in configs, this is not needed anymore. For BC, `!env_vars!` are now resolved on PHP's side, with a deprecation notice.

@johnstevenson sorry, I did not use your code - it did not fit here :)

File: UPGRADE-3.3.md [hunk 1]
```diff
@@ -42,6 +42,17 @@ HttpKernel
...
 
+Process
+-------
+
+ * Delayed expansion on Windows has been disabled.
+
+ * `!VAR!` expansion allowed by ProcessUtils::escapeArgument() is deprecated.
+   You should resolve these variables before calling `ProcessUtils::escapeArgument()`.
+
+ * Configuring `proc_open()` options, Windows compatibility and SIGCHLD
```
Comment (stof) [inline UPGRADE-3.3.md]: missing ``I`` in ``SIGCHILD`` (same in other files)

File: src/Symfony/Component/Process/Process.php [hunk 2]
```diff
@@ -171,7 +171,10 @@ public function __construct($commandline, $cwd = null, array $env = null, $input
         $this->pty = false;
         $this->enhanceWindowsCompatibility = true;
         $this->enhanceSigchildCompatibility = '\\' !== DIRECTORY_SEPARATOR && $this->isSigchildEnabled();
-        $this->options = array_replace(array('suppress_errors' => true, 'binary_pipes' => true), $options);
+        if (null !== $options) {
+            @trigger_error(sprintf('The $options parameter of the %s constructor is deprecated since version 3.3 and will be removed in 4.0.', __CLASS__), E_USER_DEPRECATED);
+            $this->options = $options + $this->options;
```
Comment (stof) [inline src/Symfony/Component/Process/Process.php]: I would keep ``array_replace``. It makes the code more readable IMO.

File: src/Symfony/Component/Process/Process.php [hunk 3]
```diff
@@ -274,24 +277,23 @@ public function start(callable $callback = null)
             if ('\\' === DIRECTORY_SEPARATOR && !empty($this->options['bypass_shell']) && !$this->enhanceWindowsCompatibility) {
                 throw new LogicException('The "bypass_shell" option must be false to inherit environment variables while enhanced Windows compatibility is off');
             }
+            $escape = '\\' === DIRECTORY_SEPARATOR ? function ($v) { return ProcessUtils::escapeArgument($v, ProcessUtils::ESC_WINDOWS_CMD); } : 'escapeshellarg';
```
Comment (stof) [inline src/Symfony/Component/Process/Process.php]: wouldn't this break BC by disabling the expansion of variables due to having 2 args ?

File: src/Symfony/Component/Process/ProcessUtils.php [hunk 4]
```diff
@@ -32,46 +34,42 @@ private function __construct()
...
+            $argument = preg_replace_callback("/!([^!=\n]++)!/", function ($m) {
+                @trigger_error(sprintf('Delayed variables are deprecated since Symfony 3.3 and will be left unresolved in 4.0. Resolve the %s variable before calling ProcessUtil::escapeArgument().', $m[0]), E_USER_DEPRECATED);
 
-            return $escapedArgument;
+                return getenv($m[1]);
+            }, $argument);
         }
 
-        return escapeshellarg($argument);
+        if ((self::ESC_WINDOWS_ARGV & $mode) && (false !== strpbrk($argument, " \t\n\v\"") || !isset($argument[0]))) {
```
Comment (stof) [inline src/Symfony/Component/Process/ProcessUtils.php]: shouldn't the case of the empty string be checked before the ``strpbrk`` ? I suspect it is a cheaper call

File: src/Symfony/Component/Process/ProcessUtils.php [hunk 5]
```diff
@@ -32,46 +34,42 @@ private function __construct()
...
+            $argument = preg_replace('/(\\\\*+)"/', '$1$1\\"', $argument);
+            $argument = preg_replace('/(\\\\++)$/', '$1$1', $argument);
```
Comment (stof) [inline src/Symfony/Component/Process/ProcessUtils.php]: for the empty string, we could skip these regexes

File: src/Symfony/Component/Process/ProcessUtils.php [hunk 6]
```diff
@@ -17,11 +17,13 @@
  * ProcessUtils is a bunch of utility methods.
  *
  * This class contains static methods only and is not meant to be instantiated.
- *
- * @author Martin Hasoň <martin.hason@gmail.com>
  */
 class ProcessUtils
 {
+    const ESC_WINDOWS_ARGV = 1;
+    const ESC_WINDOWS_CMD = 2;
```
Comment (stof) [inline src/Symfony/Component/Process/ProcessUtils.php]: we should have documentation on these to explain their respective behavior

File: src/Symfony/Component/Process/Tests/ProcessUtilsTest.php [hunk 7]
```diff
@@ -27,12 +27,12 @@ public function dataArguments()
...
-                array('^%"path"^%', '%path%'),
-                array('"<|>\\" \\"\'f"', '<|>" "\'f'),
-                array('""', ''),
-                array('"with\trailingbs\\\\"', 'with\trailingbs\\'),
+                array('^"\^"php\^" \^"-v\^"^"', '"php" "-v"'),
+                array('^"foo bar^"', 'foo bar'),
+                array('^%path^%', '%path%'),
+                array('^"^<^|^>\\^" \\^"\'f^"', '<|>" "\'f'),
+                array('^"^"', ''),
+                array('^"with\trailing bs\\\\^"', 'with\trailing bs\\'),
```
Comment (stof) [inline src/Symfony/Component/Process/Tests/ProcessUtilsTest.php]: we should also have tests covering the usage of the different constants

File: src/Symfony/Component/Process/Process.php (hunk 3 above)
Comment (stof) [inline src/Symfony/Component/Process/Process.php]: and why applying on the CMD escaping, not the ARGV escaping, while Unix uses ``escapeshellarg`` ?

**Comment by johnstevenson** [comment]
@nicolas-grekas No worries from my end for not using my code. However...

The reason I did not use the code from the blog entry was because it is not reliable (incidentally, it is listed in my [References](https://github.com/johnstevenson/winbox-args/wiki/References) section). Unfortunately, the author missed an extremely important gotcha:

> If the argument is the name of the program to execute and it contains whitespace, then it will not be recognized by _cmd_ if it is caret-escaped. This is because the program is identified from the start of the command, delimited by whitespace outside a quoted-string. And because the parser is never inside a _quoted-string_ it will stop when it reaches the first whitespace character.

from [Caret Escaping Meta Characters](https://github.com/johnstevenson/winbox-args/wiki/Caret-escaping-meta-characters) which is part of the [Implementing a solution](https://github.com/johnstevenson/winbox-args/wiki/Implementing-a-solution) section that ties everything together.

For example `C:\Program Files\php\php.exe` will be caret-escaped as `^"C:\Program Files\php\php.exe^"`. But _cmd_ will parse this as `"C:\Program` which obviously cannot be executed.

On a separate point, I do not understand the bit-masked constants, which appear to make things very brittle. For example the test

```php
if (!(self::ESC_WINDOWS_ARGV_CMD & $mode))
```

doesn't actually restrict the `$mode` argument to the range `1-3` as intended. Also, if I were to pass in `ESC_WINDOWS_CMD` then a crucial part of the escaping is ignored. Perhaps the absence of an explanation has confused me here, but surely all that is needed is a:

- default `true` we are using cmd.exe
- `false` we are not using cmd.exe

Incidentally, just before you updated the PR I was working through the appveyor test failures to check if the actual output was correct and that the test data needed changing. I am pleased to report that this is the case. Hey ho!

**Comment by nicolas-grekas** [comment]
Hum. In fact, the author of the MSDN article makes it clear to me:

> Windows knows about only one command line string for each process. Because one string is not terribly useful, libraries conspire to provide the illusion of multiple command line arguments.

`cmd` parses its parts, then for some apps, `CommandLineToArgvW` splits again the string (as do eg when calling php).
Thus, the path-to-php.exe is not parsed by `CommandLineToArgvW` - only by cmd. It doesn't take part of that "single string" that `CommandLineToArgvW` then interprets as command line arguments (the illusion mentioned above.)

All in all, that means there is no solution: escaping on Windows is contextual. If you don't know the context where the escaped value is going to be inserted, you can't escape it correctly...

Eg there is no way to escape the `\n` character.

Thus, I'm closing. Or enemy is parsing as done by cmd.

I might come up soon again with another radically different approach based on this snippet :)

```php
putenv('LF=
');
putenv('C=SET && php -r "print_r($_SERVER);" "^^^!LF^!a!LF!b!LF!c"');

proc_close(proc_open('cmd /V:ON /E:ON /D /S /C %C%', array(), $p, null, null, array('bypass_shell' => true)));
```

**Comment by johnstevenson** [comment]
>Thus, the path-to-php.exe is not parsed by CommandLineToArgvW - only by cmd.

I don't understand what you mean by this. Of course the path-to-php.exe is parsed by `CommandLineToArgvW`. See [How Windows parses the command line](https://github.com/johnstevenson/winbox-args/wiki/How-Windows-parses-the-command-line) for the details.

**Comment by nicolas-grekas** [comment]
Continued in #21254 with a radically new approach.

**Comment by nicolas-grekas** [comment]
> Of course the path-to-php.exe is parsed by CommandLineToArgvW.

Example with echo: `echo "\""` > the output is `"\""` - which means `CommandLineToArgvW` did not apply.

**Comment by johnstevenson** [comment]
Sorry, you've lost me! `echo` is an internal command so `CommandLineToArgvW` would never apply.

Just to be clear, do you believe that path-to-php.exe is **not** parsed by `CommandLineToArgvW` ?

**Comment by nicolas-grekas** [comment]
We don't care anymore, see #21254 :)

**Comment by johnstevenson** [comment]
Eh? This is #21254 isn't it?

**Comment by Simperfit** [comment]
Maybe he was talking about #21347

**Comment by nicolas-grekas** [comment]
I meant #21347 sorry :)

**Pull request closed without merging**